### ISSUE TITLE
SMTEncoding of impure arrows comes in the way of verification

### ISSUE THREAD

**Issue by aseemr** [FStarLang/FStar#855]
The following example does not verify:

```F#
val bar: unit -> St (unit -> St nat)
let bar () =
  let r = alloc 2 in
  let f:(unit -> St nat) = fun () -> !r in
  f
```

On debugging, I found that the problem is because F* encodes the arrow type `unit -> St nat` (at least) twice, and is not able to reason that it's the same type.

So, I tried the following version, and it works:

```F#
type fn = unit -> St nat

val bar: unit -> St fn
let bar () =
  let r = alloc 2 in
  let f:fn = ((fun () -> !r) <: fn) in
  f
```

Perhaps the annotations cause `unit -> St nat` to be encoded just once.

We don't encode any `HasType` axiom for impure arrows, and so this problem comes (I suspect that for pure arrows the HasType enables some reasoning of two encodings of the same type). Can we make hashconsing smarter so that it does not encode same impure arrow twice (or perhaps it already is?).

**Comment by aseemr** [comment]
We could do two things here:

1. Make hash consing of arrow types more precise, so that it takes into account the effect label, the return type, and the wp.

2. Add HasType axiom for total effectful arrows (along the lines of what we do for pure arrows).

(1) is anyway good to have for working with Div code.

**Comment by kyoDralliam** [comment]
One strange observation (at least to me), the following code does not need as much typing annotations nor let-bindings (maybe because it is done automatically in the typechecker)

```
val bar2: unit -> St (fn * unit)
let bar2 () =
  let r : ref nat = alloc 2 in
  (fun () -> !r), ()
```

**Comment by shravanrn** [comment]
Was requested by assem to post the following example here, which may be another instance of this issue

Below, I am trying to ensure that `e` and `h` always have the same length.

```
module DepStruct

module HST = FStar.HyperStack.ST
module I32 = FStar.Int32
module Seq = FStar.Seq

noeq
type wf_Q = 
  | Q_Cons :
    e : Seq.seq int ->
    h : (me:Seq.seq (int -> int -> HST.St int){ Seq.length e = Seq.length me}) ->
  wf_Q

let main() : HST.St I32.t =
  let q : wf_Q = Q_Cons Seq.empty Seq.empty in
  0l
```

But the program gives the following type error
```
DepStruct.fst(11,4-11,12): (Error 19) Subtyping check failed; expected type me:
FStar.Seq.Base.seq (_: Prims.int -> _: Prims.int -> FStar.HyperStack.ST.St Prims.int)
  {FStar.Seq.Base.length (Q_Cons?.events projectee) = FStar.Seq.Base.length me}; got type me:
FStar.Seq.Base.seq (_: Prims.int -> _: Prims.int -> FStar.HyperStack.ST.St Prims.int)
  {FStar.Seq.Base.length events = FStar.Seq.Base.length me} (see also DepStruct.fst(11,55-11,88))
DepStruct.fst(15,34-15,43): (Error 19) Subtyping check failed; expected type me:
FStar.Seq.Base.seq (_: Prims.int -> _: Prims.int -> FStar.HyperStack.ST.St Prims.int)
  {FStar.Seq.Base.length FStar.Seq.Base.empty = FStar.Seq.Base.length me}; got type s:
FStar.Seq.Base.seq (_: Prims.int -> _: Prims.int -> FStar.HyperStack.ST.St Prims.int)
  {FStar.Seq.Base.length s = 0} (see also DepStruct.fst(11,55-11,88))
Verified module: DepStruct (1225 milliseconds)
2 errors were reported (see above)
```

This does work when i change the effect from `St` to `Pure`

**Comment by aseemr** [comment]
Fixed by #2005, both the testcases work in F* master, closing.
